src/extensions/extensionsHelper.ts

```diff
@@ -506,52 +531,284 @@ export async function publishExtensionVersionFromLocalSource(args: {
+          extension.repoUri!
+        )}' already associated with Extension ${clc.bold(
+          extensionRef
+        )}. Repo URI cannot be changed.`
+      );
+    } else {
+      logger.info(
+        `Extension ${clc.bold(extensionRef)} is published from ${clc.bold(extension?.repoUri)}.`
+      );
+    }
+  } else {
+    logger.info(
+      `\n${clc.red("Warning:")} You are about to associate repo URI ${clc.bold(
+        repoUri
+      )} with Extension ${clc.bold(
+        extensionRef
+      )}. This cannot be changed. All future verifiable versions must be published from this repo. ` +
+        `You can continue publishing unverifiable versions from local source.`
+    );
+    const confirmed = await confirm({
       nonInteractive: args.nonInteractive,
       force: args.force,
       default: false,
-    }))
-  ) {
-    return;
+    });
+    if (!confirmed) {
+      return;
+    }
+  }
+
+  // Prompt for Extension root and default to last root used.
+  let extensionRoot = args.extensionRoot;
+  let defaultRoot = "/";
+  if (!extensionRoot) {
+    if (!args.nonInteractive) {
+      if (extension) {
+        try {
+          const extensionVersionRef = `${extensionRef}@${extension.latestVersion}`;
+          const extensionVersion = await getExtensionVersion(extensionVersionRef);
+          defaultRoot = extensionVersion.extensionRoot ?? defaultRoot;
+        } catch (err: any) {
+          // Not a critical error so continue with default root.
+        }
+      }
+      extensionRoot = await promptOnce({
+        type: "input",
+        message:
+          "Enter this Extension's root directory in the repo (defaults to previous root if set):",
+        default: defaultRoot,
+      });
+    } else {
+      extensionRoot = defaultRoot;
+    }
   }
 
-  if (
-    extension &&
-    extension.latestVersion &&
-    semver.lt(extensionSpec.version, extension.latestVersion)
-  ) {
-    // publisher's version is less than current latest version.
+  // Prompt for source ref and default to HEAD.
+  let sourceRef = args.sourceRef;
+  const defaultSourceRef = "HEAD";
+  if (!sourceRef) {
+    if (!args.nonInteractive) {
+      sourceRef = await promptOnce({
+        type: "input",
+        message: "Enter the commit hash, branch, or tag name to build from in the repo:",
+        default: defaultSourceRef,
+      });
+    } else {
+      sourceRef = defaultSourceRef;
+    }
+  }
+
+  // Prompt for release stage.
+  let stage = args.stage;
+  const defaultStage = "rc";
+  if (!stage) {
+    if (!args.nonInteractive) {
+      stage = await promptOnce({
+        type: "list",
+        message: "Choose the release stage (pre-release annotations will be auto-incremented):",
+        choices: stageOptions,
+        default: defaultStage,
+      });
+    } else {
+      stage = defaultStage;
+    }
+  }
+
+  // Fetch and validate Extension from remote repo.
+  const archiveUri = path.join(repoUri, `archive/${sourceRef}.zip`);
```

Comment:
path.join is only appropriate for local paths, not for URIs. This will strip the second / in `https://`, because it interprets it as an empty segment in a local path